module/documents/actor/actor.mjs

```diff
@@ -1864,78 +1881,156 @@ export default class Actor5e extends Actor {
+        }).map(e => e._id);
+
+        for (const e of oEffects) {
+          const item = e.origin?.startsWith("Actor") ? await fromUuid(e.origin) : {};
+          const originIsSelf = item.parent?.uuid === this.uuid;
+          const isOriginEffect = originEffectIds.includes(e._id);
+
+          if ( isOriginEffect ) {
+            // If effect originates on actor
+            if (!removeOriginAE) d.effects.push(e);
+          } else if ( !isOriginEffect && !originIsSelf ) {
+            // Effect is not from Actor
+            if ( !removeOtherOriginAE ) d.effects.push(e);
+          } else {
+            // Effect is from an item originating on actor
+            switch (item.type) {
+              case "spell": {
+                if ( !removeSpellAE ) d.effects.push(e);
+                break;
+              }
+              case "feat": {
+                if ( !removeFeatAE ) d.effects.push(e);
+                break;
+              }
+              case "subclass":
+              case "class": {
+                if ( !removeClassAE ) d.effects.push(e);
+                break;
+              }
+              case "equipment":
+              case "weapon":
+              case "tool":
+              case "loot":
+              case "backpack": {
+                if ( !removeEquipmentAE ) d.effects.push(e);
+                break;
+              }
+              case "background": {
+                if ( !removeBackgroundAE ) d.effects.push(e);
+                break;
+              }
+              default: {
+                // Unknown type, or origin was not caught by above filters, keep
+                d.effects.push(e);
+              }
+            }
+          }
+        }
+      }
     }
 
-    // Keep biography
-    if ( keepBio ) d.system.details.biography = o.system.details.biography;
-
-    // Keep senses
-    if ( keepVision ) d.system.traits.senses = o.system.traits.senses;
+    // Set a random image if source is configured that way
+    if ( source.prototypeToken.randomImg ) {
+      const images = await target.getTokenImages();
+      d.prototypeToken.texture.src = images[Math.floor(Math.random() * images.length)];
+    }
 
     // Set new data flags
     if ( !this.isPolymorphed || !d.flags.dnd5e.originalActor ) d.flags.dnd5e.originalActor = this.id;
     d.flags.dnd5e.isPolymorphed = true;
 
-    // Update unlinked Tokens in place since they can simply be re-dropped from the base actor
+    // Gather previous actor data
+    const previousActorIds = this.getFlag("dnd5e", "previousActorIds") || [];
+    previousActorIds.push(this._id);
+    foundry.utils.setProperty(d.flags, "dnd5e.previousActorIds", previousActorIds);
+
+    // Update unlinked Tokens, and grab a copy of any actorData adjustments to re-apply
     if ( this.isToken ) {
       const tokenData = d.prototypeToken;
       delete d.prototypeToken;
+      d.flags.dnd5e.previousActorData = duplicate(this.token.actorData);
```

Comment:
```suggestion
      d.flags.dnd5e.previousActorData = this.token.toObject().actorData;
```

And should this flag be stored on the `TokenDocument` itself rather than placed in its `actorData` override?

Reply:
I'll loop back to this, I had problems storing it on the token for reasons I need to remember.